From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wayne Chang <waynec@nvidia.com>
Date: Mon, 27 Oct 2025 02:37:59 +0000
Subject: [PATCH 142/243] NVIDIA: SAUCE: usb: xhci: Check PSIM value to
 determine USB 3.2 Gen rate

The current implementation assumes all USB 3.2 (xHCI minor_rev == 2)
hosts support Gen 2x2 (20 Gbps) operation. However, the USB 3.2
specification defines two possible speed configurations:
- USB 3.2 Gen 1x2: 10 Gbps (2 lanes @ 5 Gbps each)
- USB 3.2 Gen 2x2: 20 Gbps (2 lanes @ 10 Gbps each)

This leads to incorrect ssp_rate reporting for Gen 1x2 capable hosts.

Fix this by parsing the Protocol Speed ID (PSI) table from the xHCI
extended capabilities to determine the actual maximum speed supported
by the host controller. The PSIM (Protocol Speed ID Mantissa) field
indicates the actual speed:
- PSIM = 10 (Gbps) -> USB_SSP_GEN_1x2
- PSIM = 20 (Gbps) -> USB_SSP_GEN_2x2

If PSI information is not available (psi_count == 0), the code defaults
to USB_SSP_GEN_2x2 to maintain backward compatibility with hosts that
don't provide detailed PSI tables.

Bug 4720659

Change-Id: I3671c205e43d3fcd06ffc99a73d60c1a5570ca9e
Signed-off-by: Wayne Chang <waynec@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3476795
Reviewed-by: EJ Hsu <ejh@nvidia.com>
Reviewed-by: WK Tsai <wtsai@nvidia.com>
Reviewed-by: Henry Lin <henryl@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit c9281038797f8b5b356137a0ec8a96840735a6a3)

diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index 23b104c2956c..39527c98526f 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -5399,11 +5399,58 @@ static void xhci_hcd_init_usb3_data(struct xhci_hcd *xhci, struct usb_hcd *hcd)
 
 	switch (minor_rev) {
 	case 2:
+		struct xhci_port_cap *port_cap = NULL;
+		int i;
+
 		hcd->speed = HCD_USB32;
 		hcd->self.root_hub->speed = USB_SPEED_SUPER_PLUS;
 		hcd->self.root_hub->rx_lanes = 2;
 		hcd->self.root_hub->tx_lanes = 2;
+		/*
+		 * Determine the SSP rate based on the PSIM value from PSI table.
+		 * USB 3.2 can be either Gen 1x2 (10 Gbps) or Gen 2x2 (20 Gbps).
+		 * Default to Gen 2x2 if PSI information is not available.
+		 */
 		hcd->self.root_hub->ssp_rate = USB_SSP_GEN_2x2;
+
+		/* Find the USB 3.x port capability */
+		for (i = 0; i < xhci->num_port_caps; i++) {
+			if (xhci->port_caps[i].maj_rev == 0x03) {
+				port_cap = &xhci->port_caps[i];
+				break;
+			}
+		}
+
+		if (port_cap && port_cap->psi_count) {
+			u16 max_rate = 0;
+			u8 psie;
+			u16 psim;
+
+			/* Find the maximum speed from PSI entries */
+			for (i = 0; i < port_cap->psi_count; i++) {
+				psie = XHCI_EXT_PORT_PSIE(port_cap->psi[i]);
+				psim = XHCI_EXT_PORT_PSIM(port_cap->psi[i]);
+
+				/* Convert to Gbps */
+				while (psie < 3) { /* 3 = Gbps */
+					psim /= 1000;
+					psie++;
+				}
+
+				if (psim > max_rate)
+					max_rate = psim;
+			}
+
+			/*
+			 * Set SSP rate based on maximum speed:
+			 * 10 Gbps = USB 3.2 Gen 1x2 (2 lanes @ 5 Gbps each)
+			 * 20 Gbps = USB 3.2 Gen 2x2 (2 lanes @ 10 Gbps each)
+			 */
+			if (max_rate == 10)
+				hcd->self.root_hub->ssp_rate = USB_SSP_GEN_1x2;
+			else if (max_rate == 20)
+				hcd->self.root_hub->ssp_rate = USB_SSP_GEN_2x2;
+		}
 		break;
 	case 1:
 		hcd->speed = HCD_USB31;
-- 
2.43.0

